psi4/driver/driver_cbs.py

```diff
@@ -701,13 +936,170 @@ def return_energy_components():
     return VARH
 
+
 VARH = return_energy_components()
 
 
+def _get_default_xtpl(nbasis, xtpl_type):
+    """ A helper function to determine default extrapolation type.
+
+    Parameters
+    ----------
+    nbasis : int
+        Number of basis sets
+    xtpl_type : {'scf', 'corl'}
+        Extrapolation type: 'scf' for the total energy, 'corl' for just the
+        correlation component.
+
+    Returns
+    -------
+    function
+        Extrapolation function to be used.
+    """
+
+    if nbasis == 1:
+        return xtpl_highest_1
+    elif xtpl_type == "scf":
+        if nbasis == 2:
+            return scf_xtpl_helgaker_2
+        elif nbasis == 3:
+            return scf_xtpl_helgaker_3
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to scf_xtpl: %d" % nbasis)
+    elif xtpl_type == "corl":
+        if nbasis == 2:
+            return corl_xtpl_helgaker_2
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to corl_xtpl: %d" % nbasis)
+
+
+def _validate_cbs_inputs(cbs_metadata):
+    """ A helper function which validates the ``cbs_metadata`` format,
+    expands basis sets, and provides sensible defaults for optional arguments.
+
+    Parameters
+    ----------
+    cbs_metadata : array
+        array of dicts containing CBS stage keywords.
+
+    Returns
+    -------
+    array
+        Array of dictionaries, with each item consisting of an extrapolation
+        stage. All validation takes place here.
+    """
+
+    metadata = []
+    for item in cbs_metadata:
+        # 1a) all items must have wfn
+        if "wfn" not in item:
+            raise ValidationError("Stage {d} doesn't have defined level of theory!".format(cbs_metadata.index(item)))
+    # 1b) all items must have basis set
+        if "basis" not in item:
+            raise ValidationError("Stage {d} doesn't have defined basis sets!".format(cbs_metadata.index(item)))
+    # 2a) process required stage parameters and assign defaults
+        stage = {}
+        stage["wfn"] = item["wfn"].lower()
+        stage["basis"] = _expand_bracketed_basis(item["basis"].lower())
+        # 2b) if first item is not HF, generate it
+        if len(metadata) == 0 and stage["wfn"] not in ["hf", "c4-hf", "scf", "c4-scf"]:
+            scf = {}
+            if stage["wfn"].startswith("c4"):
+                scf["wfn"] = "c4-hf"
+            else:
+                scf["wfn"] = "hf"
+            scf["basis"] = ([stage["basis"][0][-1]], [stage["basis"][1][-1]])
+            scf["treatment"] = "scf"
+            scf["stage"] = "scf"
+            scf["scheme"] = _get_default_xtpl(len(scf["basis"][1]), scf["treatment"])
+            scf["alpha"] = None
+            scf["options"] = False
+            metadata.append(scf)
+    # 2c) keep processing current stage
+        stage["treatment"] = item.get("treatment", "scf" if len(metadata) == 0 else "corl")
+        stage["stage"] = item.get("stage", False)
```

Comment:
I agree on `stage` -- it'll be nice to have variable labels. Is that in a test somewhere?

Not so sure about `treatment`. It's switching btwn `scf` and `corl`, right? I wouldn't have expected past code to handle a non-scf-or-corl value correctly. Great if it can now, just not seeing it from a quick search.

Reply:
Because of GitHub's infernal new "helpfullness" in hiding conversations, I can't find the `basis_lo` thread. I understand your use-case now and agree. And perhaps if there were several alternate formulas to be applied to `"AE - FC"` then that would constitute a new `treatment` class in your view? For `basis_lo`, since that's a new feature, I'd feel better if it had a test to make sure we don't break it.

Reply:
In the past code, you could only ever pass in one `scf` stage. The new code, due to `_get_default_xtpl` requires each stage to have either `corl` or `scf` set using the `treatment` tag. One can, in principle, pass two `scf` stages using the new code. I can't think of why someone would want to do that, perhaps with the exception of relativistic corrections over scf values from extrapolated basis sets. It's a very niche use case (the xtpl formula for scf and corl for 1 basis set is the same - xtpl_highest_1). I can of course have the code ignore user input.

There is no test for `stage`; it is really just a print label - I didn't think it was necessary.

The `AE - FC` class is still correction (or extrapolation with 2 basis sets) of correlation; it doesn't require a new class of `treatment`.

I'll add a test for `basis_lo`.

Reply:
Thanks. I think the `basis_lo` is the important test. Throw in an alternate `stage` label, just so the line of code gets hit and we may notice if something breaks.

For `treatment`, pass `treatment='leeches'` and if a sensible error happens, I'm good with having it user-settable. If error is obscure, best to take that knob away from the user. (If calculation works, I'm backing away slowly.)

Reply:
Leeches now throw a validation error ;)